File: day008_string_traversal/related_letter_coverage/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cctype>

using namespace std;

// Optimized: O(N) Time, O(1) Space - Use a fixed-size boolean array (hashing)
void solve_letter_coverage() {
    string S;
    if (!getline(cin, S)) return;

    // Fixed size array for a-z (O(1) space)
    vector<bool> seen(26, false);
    int unique_count = 0;

    for (char c : S) {
        char lower_c = tolower(c);
        
        if (lower_c >= 'a' && lower_c <= 'z') {
            int index = lower_c - 'a';
            
            if (!seen[index]) {
                seen[index] = true;
                unique_count++;
            }
        }
    }
    
    if (unique_count == 26) {
        cout << "Pangram: YES" << endl;
    } else {
        cout << "Pangram: NO" << endl;
    }
}

void solve_letter_coverage_brute() {
    // Brute: Checking each letter 'a' through 'z' individually using string::find (O(N*26) approx O(N))
    solve_letter_coverage(); 
}

int main() {
    solve_letter_coverage();
    return 0;
}

File: day008_string_traversal/related_letter_coverage/solution.java
import java.util.Scanner;
import java.util.Arrays;

class Solution {

    // Optimized: O(N) Time, O(1) Space - Boolean Array Hashing
    public static void solve_letter_coverage() {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String S = scanner.nextLine().toLowerCase();
        
        // Fixed size array for a-z (O(1) space)
        boolean[] seen = new boolean[26];
        int uniqueCount = 0;

        for (int i = 0; i < S.length(); i++) {
            char c = S.charAt(i);
            
            if (c >= 'a' && c <= 'z') {
                int index = c - 'a';
                
                if (!seen[index]) {
                    seen[index] = true;
                    uniqueCount++;
                }
            }
        }
        
        if (uniqueCount == 26) {
            System.out.println("Pangram: YES");
        } else {
            System.out.println("Pangram: NO");
        }
        scanner.close();
    }

    public static void main(String[] args) {
        solve_letter_coverage();
    }
}